### DISCUSSION TITLE
Append extra data in FormData

### DISCUSSION THREAD

**Discussion by NormandoHall** [sveltejs/kit#7581] [Q&A] [answered]
I have a form and I want to add reCaptcha `token`, but this token is generated when the user click the submit button. I can't await for reCaptcha, and of course I can't append the token data. This is my code for `reCaptcha V3`

```svelte
  <form
    class="space-y-6"
    action="?/login"
    method="POST"
    use:enhance={({ data }) => {
      console.log('1');

      grecaptcha.ready(function () {
        grecaptcha.execute(PUBLIC_RECAPTCHA_SITE_KEY, { action: 'submit' }).then(function (t) {
          console.log('2');
          data.append('token', t);
        });
      });

      console.log('3');

      return async ({ result }) => {
        await invalidateAll();
        await applyAction(result);
      };
    }}
  >

    <input..../>
    <input.../>

    <button type="submit">Submit>/button>
  </form>
```

I get console log 1, 3, 2.

Thanks

**Comment by teemingc** [comment, marked as answer]
Does this work?
```diff
  <form
    class="space-y-6"
    action="?/login"
    method="POST"
-    use:enhance={({ data }) => {
+   use:enhance={ async ({data}) => {
      console.log('1');

+  const captcha = new Promise((resolve) => {
      grecaptcha.ready(function () {
        grecaptcha.execute(PUBLIC_RECAPTCHA_SITE_KEY, { action: 'submit' }).then(function (t) {
          console.log('2');
          data.append('token', t);
+        resolve();
        });
      });
+   });
+   await captcha;
      console.log('3');

      return async ({ result }) => {
        await invalidateAll();
        await applyAction(result);
      };
    }}
  >

    <input..../>
    <input.../>

    <button type="submit">Submit>/button>
  </form>
```

**Reply by NormandoHall** [reply]
Thanks @s3812497 for the reply. I was tested something similar and I think I was make a mistake, but get the same result. This is the output at dev tools:

![image](https://user-images.githubusercontent.com/2943127/201108117-a47efe91-75ab-4968-b439-0d90f3c99f62.png)

If I check the error link (form.js....):

![image](https://user-images.githubusercontent.com/2943127/201108367-80e19b7c-b344-42ea-93fd-8504c711105a.png)

**Reply by NormandoHall** [reply]
Errors comes when make the enhance function as async.

This give me an error

```svelte
  <form
    class="space-y-6"
    action="?/login"
    method="POST"
    use:enhance={(async { data }) => { // <-- async

      return async ({ result }) => {
        await invalidateAll();
        await applyAction(result);
      };
    }}
  >

    <input..../>
    <input.../>

    <button type="submit">Submit>/button>
  </form>
```

This is not:

```svelte
  <form
    class="space-y-6"
    action="?/login"
    method="POST"
    use:enhance={({ data }) => {

      return async ({ result }) => {
        await invalidateAll();
        await applyAction(result);
      };
    }}
  >

    <input..../>
    <input.../>

    <button type="submit">Submit>/button>
  </form>
```

**Reply by NormandoHall** [reply]
Will it be necessary to create an issue for this?

**Reply by teemingc** [reply]
> Errors comes when make the enhance function as async.
> 
> 
> 
> This give me an error
> 
> 
> 
> ```svelte
> 
>   <form
> 
>     class="space-y-6"
> 
>     action="?/login"
> 
>     method="POST"
> 
>     use:enhance={(async { data }) => { // <-- async
> 
> 
> 
>       return async ({ result }) => {
> 
>         await invalidateAll();
> 
>         await applyAction(result);
> 
>       };
> 
>     }}
> 
>   >
> 
> 
> 
>     <input..../>
> 
>     <input.../>
> 
> 
> 
>     <button type="submit">Submit>/button>
> 
>   </form>
> 
> ```

The async keyword is in the wrong place for the enhance action.
```diff
- (async {data}) => {}
+ async ({data}) => {}
```

**Reply by NormandoHall** [reply]
OMG! You are right, I make a mistake when copy the code. But still get the same error. 
You can test it only appending the `async` keyword to any enhanced form

`use:enhance={async ({ data }) => {`

**Reply by teemingc** [reply]
sorry about that. You're right, the callback cannot be an async function.

I finally got down to writing and testing a solution.
Maybe something like this might help?
https://stackblitz.com/edit/sveltejs-kit-template-default-xk6efm?file=src%2Froutes%2F%2Bpage.svelte&terminal=dev

```svelte
<script>
  import { enhance } from '$app/forms'
</script>

<form method="POST" use:enhance={({data}) => {
  alert('Start');

	const x = new Promise((resolve) => {
		setTimeout(() => {
			alert('callback');
			resolve();
		}, 1000);
	});

	return async () => {
		await x;
		alert('Finish')
	}
}}>
<p>Press submit to see the order of alerts.</p>
<button>Submit</button>
</form>
```

**Reply by NormandoHall** [reply]
Well, I tested, and runs in sequence correctly, but I can't add the extra parameter because when you call `await x` you are in the return function, after send the form, not before send.

**Reply by NormandoHall** [reply]
So, I can workaround this issue of course, using a legacy endpoint API `server.js`, but the idea is to use enhanced forms. I think this is an issue, because we can't stop or delay the form sending, only `cancel()`.

**Reply by teemingc** [reply]
https://kit.svelte.dev/docs/form-actions#progressive-enhancement-custom-event-listener

This might be the best solution for you. You can still use form actions, but you have to define your own "enhance" function.
This gives you more fine-grained control of when the form is submitted. Also note that the function is async now.

```html
<script lang="ts">
  import { invalidateAll, goto } from '$app/navigation';
  import { applyAction } from '$app/forms';

  import type { ActionData } from './$types';
  import type { [ActionResult](https://kit.svelte.dev/docs/types#sveltejs-kit-actionresult) } from '@sveltejs/kit';

  export let form: ActionData;

  let error: any;

  async function handleSubmit(event) {
    const data = new FormData(this);

    // await your recaptcha stuff here and append token to form data.

    const response = await fetch(this.action, {
      method: 'POST',
      body: data
    });

    const result: [ActionResult](https://kit.svelte.dev/docs/types#sveltejs-kit-actionresult) = await response.json();

    if (result.type === 'success') {
      // re-run all `load` functions, following the successful update
      await invalidateAll();
    }

    applyAction(result);
  }
</script>

<form method="POST" on:submit|preventDefault={handleSubmit}>
  <!-- content -->
</form>
```

**Reply by NormandoHall** [reply]
Thanks @s3812497 Tee. Runs ok. 

I think must be described in documentation about it can't be async, for clarification to newcomers. Or better, one day make it async! xD

Regards!

**Reply by teemingc** [reply]
@NormandoHall thanks to your discovery, managed to submit a PR so that async functions are now allowed for `enhance`
